lib/trino-parquet/src/main/java/io/trino/parquet/reader/flat/FilteredRowRangesIterator.java

```diff
@@ -0,0 +1,193 @@
+ */
+package io.trino.parquet.reader.flat;
+
+import io.trino.parquet.reader.FilteredRowRanges;
+
+import java.util.Iterator;
+import java.util.OptionalLong;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static io.trino.parquet.reader.FilteredRowRanges.RowRange;
+import static java.lang.Math.min;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * When filtering using column indexes we might skip reading some pages for different columns. Because the rows are
+ * not aligned between the pages of the different columns it might be required to skip some values. The values (and the
+ * related rl and dl) are skipped based on the iterator of the required row indexes and the first row index of each
+ * page.
+ * For example:
+ *
+ * <pre>
+ * rows   col1   col2   col3
+ *      ┌──────┬──────┬──────┐
+ *   0  │  p0  │      │      │
+ *      ╞══════╡  p0  │  p0  │
+ *  20  │ p1(X)│------│------│
+ *      ╞══════╪══════╡      │
+ *  40  │ p2(X)│      │------│
+ *      ╞══════╡ p1(X)╞══════╡
+ *  60  │ p3(X)│      │------│
+ *      ╞══════╪══════╡      │
+ *  80  │  p4  │      │  p1  │
+ *      ╞══════╡  p2  │      │
+ * 100  │  p5  │      │      │
+ *      └──────┴──────┴──────┘
+ * </pre>
+ *
+ * The pages 1, 2, 3 in col1 are skipped, so we have to skip the rows [20, 79]. Because page 1 in col2 contains values
+ * only for the rows [40, 79] we skip this entire page as well. To synchronize the row reading we have to skip the
+ * values (and the related rl and dl) for the rows [20, 39] in the end of the page 0 for col2. Similarly, we have to
+ * skip values while reading page0 and page1 for col3.
+ */
+public class FilteredRowRangesIterator
+        implements RowRangesIterator
+{
+    // Iterator over all row ranges, only not-null if column index is present
+    private final Iterator<RowRange> rowRangeIterator;
+
+    // The current row range
+    private RowRange currentRange;
+
+    private long pageFirstRowIndex = Long.MAX_VALUE;
+    private int pageValuesConsumed;
+
+    public FilteredRowRangesIterator(FilteredRowRanges rowRanges)
+    {
+        requireNonNull(rowRanges, "rowRanges is null");
+        this.rowRangeIterator = rowRanges.rowRanges().iterator();
+        // If rowRanges is empty then all pages would be eliminated by filterOffsetIndex
+        checkArgument(this.rowRangeIterator.hasNext(), "rowRanges is empty");
+        nextRange();
+    }
+
+    /**
+     * @return Size of the next read within current range, bounded by chunkSize.
+     * When all the rows of the current range have been read, advance to the next range.
+     */
+    @Override
+    public int advanceRange(int chunkSize)
+    {
+        long rangeEnd = currentRange.end();
+        int rowsLeftInRange = toIntExact(rangeEnd - pageFirstRowIndex) - pageValuesConsumed + 1;
```

Comment:
`pageFirstRowIndex` is initialized to Long.MAX_VALUE. If this method gets called first, `rowsLeftInRange` will be negative -- likely a bug. In general, the contract for this class seems non-trivial, so it should have some `checkState` invocations to ensure methods are called only when appropriate.

Reply:
I've changed this to initialise `pageFirstRowIndex` to -1 and added checks to all methods to ensure that it is initialised to a valid value before it is used.